backend/de.metas.manufacturing/src/main/java/de/metas/manufacturing/generatedcomponents/MacAddressGenerator.java

```diff
@@ -0,0 +1,179 @@
+import org.adempiere.mm.attributes.api.AttributeConstants;
+import org.adempiere.mm.attributes.api.ImmutableAttributeSet;
+import org.compiere.SpringContextHolder;
+import org.compiere.util.Env;
+import org.springframework.stereotype.Service;
+
+import static de.metas.manufacturing.generatedcomponents.MacAddress.GROUP_DELIMITER;
+
+/**
+ * A MAC Address looks like this: 01:23:45:67:89:AB
+ */
+@Service
+public class MacAddressGenerator implements IComponentGenerator
+{
+	/*package*/ static final String PARAM_MAC_ADDRESS_AD_SEQUENCE_ID = "MacAddress_AD_Sequence_ID";
+	private static final int NUMBER_OF_DIGITS = 12;
+
+	private final ImmutableList<AttributeCode> supportedAttributes = ImmutableList.of(
+			AttributeConstants.RouterMAC1,
+			AttributeConstants.RouterMAC2,
+			AttributeConstants.RouterMAC3,
+			AttributeConstants.RouterMAC4,
+			AttributeConstants.RouterMAC5,
+			AttributeConstants.RouterMAC6
+	);
+
+	private transient final IDocumentNoBuilderFactory documentNoBuilder = SpringContextHolder.instance.getBean(IDocumentNoBuilderFactory.class);
+
+	@Override
+	public ImmutableAttributeSet generate(final int qty, final @NonNull ComponentGeneratorParam parameters, final @NonNull ImmutableAttributeSet existingAttributes)
+	{
+		Check.errorIf(qty < 1 || qty > 6, "Qty of Mac Addresses should be between 1 and 6. Requested qty: {}", qty);
+
+		final ImmutableList<AttributeCode> attributesToGenerate = computeRemainingAttributesToGenerate(existingAttributes);
+
+		if (attributesToGenerate.isEmpty())
+		{
+			return ImmutableAttributeSet.EMPTY;
+		}
+
+		/*
+			Explanation: we have to generate 2 mac addresses, but 2 or more are already generated => nothing to do.
+		 */
+		if (qty - noOfAlreadyGenerated(attributesToGenerate) <= 0)
+		{
+			return ImmutableAttributeSet.EMPTY;
+		}
+
+		final DocSequenceId sequenceId = DocSequenceId.ofRepoIdOrNull(StringUtils.toIntegerOrZero(parameters.getValue(PARAM_MAC_ADDRESS_AD_SEQUENCE_ID)));
+		if (sequenceId == null)
+		{
+			throw new AdempiereException("Mandatory parameter " + PARAM_MAC_ADDRESS_AD_SEQUENCE_ID + " has invalid value.");
+		}
+
+		final ImmutableAttributeSet.Builder result = ImmutableAttributeSet.builder();
+
+		for (int i = 0; i < qty - noOfAlreadyGenerated(attributesToGenerate); i++)
+		{
+			final AttributeCode attributeCode = attributesToGenerate.get(i);
+			final MacAddress macAddress = generateNextMacAddress(sequenceId);
+			result.attributeValue(attributeCode, macAddress.getAddress());
+		}
+
+		return result.build();
+	}
+
+	private int noOfAlreadyGenerated(@NonNull final ImmutableList<AttributeCode> attributesToGenerate)
+	{
+		return getSupportedAttributes().size() - attributesToGenerate.size();
+	}
+
+	@Override
+	public ImmutableList<AttributeCode> getSupportedAttributes()
+	{
+		return supportedAttributes;
+	}
+
+	@Override
+	public ImmutableMap<String, String> getDefaultParameters()
+	{
+		return ImmutableMap.<String, String>builder()
```

Comment:
hardcoded things shall be declared as class constants

Reply:
done